test/sync/src/androidTest/kotlin/io/realm/test/mongodb/shared/SubscriptionSetTests.kt

```diff
@@ -0,0 +1,321 @@
+        subscriptions.waitForSynchronization()
+        assertEquals(SubscriptionSetState.COMPLETE, subscriptions.state)
+        subscriptions.update {
+            // `age` is not a queriable field
+            realm.query<FlexParentObject>("age > 42").subscribe("test2")
+        }
+        assertFailsWith<FlexibleSyncQueryException> {
+            subscriptions.waitForSynchronization()
+        }
+        assertEquals(SubscriptionSetState.ERROR, subscriptions.state)
+    }
+
+    @Test
+    fun size() = runBlocking {
+        val subscriptions = realm.subscriptions
+        assertEquals(0, subscriptions.size)
+        subscriptions.update {
+            realm.query<FlexParentObject>().subscribe()
+        }
+        assertEquals(1, subscriptions.size)
+        subscriptions.update {
+            removeAll()
+        }
+        assertEquals(0, subscriptions.size)
+    }
+
+    @Test
+    fun errorMessage() = runBlocking {
+        val subscriptions = realm.subscriptions
+        assertNull(subscriptions.errorMessage)
+        subscriptions.update {
+            realm.query<FlexParentObject>("age > 42").subscribe()
+        }
+        assertFailsWith<FlexibleSyncQueryException> {
+            subscriptions.waitForSynchronization()
+        }
+        assertTrue(subscriptions.errorMessage!!.contains("Client provided query with bad syntax"))
+        subscriptions.update {
+            removeAll() // Removing all queries seems to provoke an error on the server, so create new valid query.
+            realm.query<FlexParentObject>().subscribe()
+        }
+        subscriptions.waitForSynchronization()
+        assertNull(subscriptions.errorMessage)
+    }
+
+    @Test
+    fun iterator_zeroSize() {
+        val subscriptions = realm.subscriptions
+        val iterator: Iterator<Subscription> = subscriptions.iterator()
+        assertFalse(iterator.hasNext())
+        assertFailsWith<NoSuchElementException> { iterator.next() }
+    }
+
+    @Test
+    fun iterator() = runBlocking {
+        val subscriptions = realm.subscriptions
+        subscriptions.update {
+            realm.query<FlexParentObject>().subscribe("sub1")
+        }
+        val iterator: Iterator<Subscription> = subscriptions.iterator()
+        assertTrue(iterator.hasNext())
+        assertEquals("sub1", iterator.next().name)
+        assertFalse(iterator.hasNext())
+        assertFailsWith<NoSuchElementException> { iterator.next() }
+        Unit
+    }
+
+    @Test
+    fun subscriptions_accessAfterRealmClosed() = runBlocking {
+        val subscriptions = realm.subscriptions
+        realm.close()
+        assertFailsWith<IllegalStateException> {
+            subscriptions.update { /* Do nothing */ }
+        }
+        Unit
+    }
+
+    @Test
+    fun waitForSynchronizationInitialSubscriptions() = runBlocking {
+        val subscriptions = realm.subscriptions
+        assertTrue(subscriptions.waitForSynchronization())
```

Comment:
Yes, before, the state was always `PENDING` but Core has changed semantics since, as that caused other issues.  So now there is a race condition whether or not this is already `COMPLETE` when we call `waitForSynchronization`